### ISSUE TITLE
AGS4 and the death of legacy

### ISSUE THREAD

**Issue by AlanDrake** [adventuregamestudio/ags#403]
So, to make coding AGS less aggravating, I'll be starting a branch whose purpose is to wipe out legacy stuff.
Whether or not another engine will steal its throne in the far future, this can be done now. 
Deleting all that stuff is certainly more pleasing than adding features wrestling with all these chains littering the source. And should make things easier to understand which in turns could also help the other AGS-inspired engines.

So far I'm experimenting and taking notes, since my knowledge of the inner workings is not as advanced as Crimson's.

I can say though that I've succesfully blasted away those pesky pixel_size/multiply_up/divide_down conversions, which had almost 400 occurences all around, and after compiling it all worked the first try. I feel greatly encouraged.

I'll create a proper branch on my repo as soon as the VS2015 branch is officially merged with master.

Things to cut away:
- legacy options
- deprecated script commands
- old audio system
- old savegames handling
- backward compatiblities to make older games run
- CD Audio
- legacy global variables (not to confuse with the new ones, distinction is not clear in the code)

**Comment by ivan-mogilko** [comment]
If you are going that way, then I strongly advise to scrap plugin API, or at least the parts of it that:
* acquire memory addresses of internal engine's objects (characters, room objects and so on)
* deal with script function pointers (get ones and register ones).
The latter normally should be remade to work same way as Script VM works inside the engine, that is - by the use of function and argument definitions.

Note, that better is just scrap all of it, and then remake useful parts, because even these are sometimes lacking. For example, when plugin implements certain engine elements, like font renderer, it should be able to tell which version of interface it supports (currently there is no way to tell, so no way to safely expand such interface).

**Comment by AlanDrake** [comment]
I'm not familiar with the plugin API. I'll see what I can do when (if) I get there. Most likely some plugin developer will have to step in, I never tried making one so I don't feel qualified enough to tamper with it.
I'll just keep cutting until things start to make sense.

**Comment by gurok** [comment]
You could probably remove the legacy pathfinding too. It will break an old Ben Jordan game, but I would rather have a known algorithm (Djikstra / A*) that we can adjust.

**Comment by ivan-mogilko** [comment]
I have a question, because I do not exactly understand the plan here.

Are you going to eventually merge those breaking changes to master branch?

If so, you need to consider that there will be two separate programs: the that runs games of certain range of versions, and another that runs games of another range of versions.

If someone would like to add a fix for the old version, or even support whole new OS to run old games on, you would have to point them out to the correct branch. Same goes for bug reports, etc.

Or someone may build the latest version because of habit, and then be surprised by old games not working with it. Perhaps the engine should display a message, suggesting them to use older version.

BTW, have you considered creating a clone of repository instead?

**Comment by ivan-mogilko** [comment]
Then again, perhaps we could let ScummVM team take the old games support on from here. Alyssa Milburn might still have her Ags for ScummVM port, which was unfinished last time I checked, but maybe someone would like to update and complete it.
I also heard they had plans to add hardware-accelerated renderer support couple of years ago, which would suit running latest AGS games.

**Comment by AlanDrake** [comment]
I really don't know... on one hand even CJ used to break compats to older versions, so it would make sense to leave behind the current backward compat version on its own branch.
On the other hand they'll end up having two different purposes, so two repos could also make sense. 
Might even be the other way around, after all the backward compat engine doesn't need the editor, it could be on its own on a separate repo.

**Comment by ivan-mogilko** [comment]
> Might even be the other way around, after all the backward compat engine doesn't need the editor, it could be on its own on a separate repo.

That too is an option, yes.

**Comment by ivan-mogilko** [comment]
> Might even be the other way around, after all the backward compat engine doesn't need the editor, it could be on its own on a separate repo.
That is another option, yes.

I am not so sure about not needing an editor, because someone might want to try restoring old game project.

**Comment by AlanDrake** [comment]
Well, I couldn't contain my excitement, so I ended up making a branch out of the pull request.
I'll omit the dll commit when I recreate the branch, if you ended up removing them.

I'm learning more about the code by taking it apart than all my previous attempts at studying it. 
I'll keep cutting away and limiting the compatibility to 341, at least until I reach the barebones stuff that is actually being used.
Then I might try some heavier refactoring if I am still alive.

**Comment by ivan-mogilko** [comment]
You may actually take a look at "experimental" branch in this repository, it contains refactored and renamed room and game classes (and many others).

**Comment by ivan-mogilko** [comment]
Another thing that I feel a need to mention, is that making a giant branches may lead to problems in the future. I've been there, so I can tell...

Common case is when you have a branch which changes 100 things at once, and you keep it for several months, then someone makes pull request that changes a code, which was also changed by you somewhere in the middle of your branch. And then you get into dilemma: either force that another contributor to wait until you finish your giant work, or merge their request now but then go back and recreate half of your branch, because it is no longer consistent.
The longer you keep unmerged branch, the more this will repeat.

**Comment by ivan-mogilko** [comment]
At last but not least... having no particular plan set for the project/team may lead to situation when you are making something in your own branch, and someone else wants to make changes to master branch, which are inconsistent, or even contradicting to yours. I just think I have to point that out, because the opinion, for example, on what to keep and what to strip out, may vary between collaborators.

The lack of strictly defined plan may also cause some people wanting to work in master branch, unaware of future break of compatibility (assuming this is what going to happen according to your idea), which would also take them by surprise.

**Comment by AlanDrake** [comment]
My current plan is to simply to strip down 3.4.1 from dead features and backwards compatibility, then do some refactoring after having learned the big picture.
When I have a clean 3.4.1 compliant AGS we may see about the future.

If possible I would force (our few) collaborators to keep their current work on a separated 3.4.1 branch (name it 3.4.1-classic or something).
Or they can just roll with my butchered branch, I'll be keeping compatibility with 3.4.1 gamedata for a long time, and in the plus side they don't have to deal with arcane stuff like those coord multipliers that make your brain melt.

We can't just keep delaying forever because people change the code with new features. You should have indeed merged your refactored classes back then and forced everyone to deal with it. (I will try to take inspiration from your experimental branch if I can)

Otherwise, if we need to start a new repository, we could announce officially that the current ags repo no longer accepts new features, except fixes and improvements.

If collaborators want to voice their concerns about their features then they should speak up, else we can just assume their tacit approval. 😁

**Comment by ivan-mogilko** [comment]
Alright, I will just leave this be then.

**Comment by ivan-mogilko** [comment]
Something that just came to my mind (as I made a big refactoring PR): dropping compatibility will reduce the range of games you can use to test the engine.
I heard you are developing a game yourself, so it may be easier in your case, but that may make sense to create a collection of test games.

**Comment by AlanDrake** [comment]
You're right.
If we had a tutorial game that implements one feature for each room, it would make both our testing easier and the users happier. 

This could be a community effort, make people pick a feature, prepare rooms then compile them into a big project.
I would have really really loved to have such a thing when I first tried AGS. The demo game is a poor reference. 
What do we do ? Open a new issue to talk about this or just post on the forum activities to get the ball rolling?

**Comment by ivan-mogilko** [comment]
I suggested a sample game for the showcase, which is a bit different, but someone mentioned feature per room idea too.

**Comment by AlanDrake** [comment]
Just so you know, I've broken the part that handles upscaling when using lowres assets in hires mode.
Which is probably good, I'll just remove it completely for the time being, I'm sure it can be rewritten in a much better and cleaner way.
Also it would be nice if one could specify the scale instead of just High/Low on the Sprites manager. this needs some thinking over.
I probably could simply upscale the assets when I start the engine, but we need to decide how to treat the sprite when used by DrawingSurfaces. Use the original or the upscaled one ? Both cases have their pros and cons.

**Comment by ivan-mogilko** [comment]
I do not know if I am expected to answer anything?
Since there is still no clear plan stated yet, just a general direction, and you are working alone in your own branch, I frankly do not see a point of discussing such things here.

**Comment by AlanDrake** [comment]
It's just my attempt to involve our dormant team. After all the whole purpose of this is to make AGS more pleasant to code for us and newcomers.
I'll keep an ear open for ideas, and I'll also keep pestering everyone so they remain aware this branch exists, otherwise it'll just end up like all the other dead branches.

Is there anything else from your branches that I could try bringing in ?

**Comment by ivan-mogilko** [comment]
This all confuses me so much, that I hardly know what to reply.

Is this a branch where you work alone, gather ideas from everyone but implement them yourself?
Are you planning to keep that branch in your repository only? If not, how long since you want to merge to this one, or rather, what point do you want to reach before doing so?

Indefinite goals and ending conditions - that is what makes branches die.

**Comment by AlanDrake** [comment]
> Are you planning to keep that branch in your repository only? If not, how long since you want to merge to this one, or rather, what point do you want to reach before doing so?

No, but I think I'll need at least another couple of weeks before reaching a stable point, then I'll try to make a pull request. A lot of things are finally starting to make sense to me, but I'm still stumbling around half blind. I'd rather not work alone, but at the moment it doesn't look I get to choose.

> Are you planning to keep that branch in your repository only? If not, how long since you want to merge to this one, or rather, what point do you want to reach before doing so?

I'll be sure to ask the forums and discord/irc as well.

**Comment by ivan-mogilko** [comment]
I asked, because from your previous question (about bringing something from my branches) it sounded like you are looking to put lots of unrelated things in your branch. I did similar things before and that only ended up with trouble.

That's why we've made a [branch/commit convention](http://www.adventuregamestudio.co.uk/forums/index.php?topic=50001.0), to keep things organized, which suggests that feature branches are dedicated to strictly defined tasks. That really helped to keep things organized.

**Comment by ivan-mogilko** [comment]
What I should probably mention too; if everyone have decided to follow compatibility breaking path, then you do not have to do whole "legacy cleanup" thing (whatever you put into the definition of "legacy") in one go. If things are not easy/clear for you, instead you could choose particular legacy feature you consider worth getting rid of, and make a pull request. And then continue doing that step by step. This will also let other people to participate (for example, I've seen @gurok mentioning changes of pathfinder).

**Comment by AlanDrake** [comment]
You're right again, there's some stuff I could already prepare a PR for. Some other instead is proving to be a little more problematic, by taking out the multipliers I have opened a whole can of worms, I'm afraid it'll take several separate attempts to get it straight, of course it had to be the first thing I tried to do...
I'll keep using my branch as a testing ground, I wish I could add comments to code without actually modifying the file, I'm littering the code with my naive observations but it can't be helped I need them...
I'll prepare the branches once I'm reasonably sure they're safe.

**Comment by ivan-mogilko** [comment]
>  I wish I could add comments to code without actually modifying the file, I'm littering the code with my naive observations but it can't be helped I need them...

That's not a problem at all, I've been adding similar notes too when I just started. Any comments in code is better than none.

**Comment by ivan-mogilko** [comment]
@AlanDrake , I tried to hold myself for some time, since I honestly wanted to cease my participation on this project, but probably I still should have said this... I apologize I did not do this earlier, but I was not in proper mental state, and I did not want to argue. And I am not sure if what I am going to say holds any value in current circumstances.

The thing is that I do not agree with what you are planning to do.

Indeed that's partially because you are going to throw away all the functionality to run old games, which we were gathering for years, piece by piece, spending so much time just to let people play them on contemporary platforms. What would happen to that now, if, for example, the old branch won't run on some newer version of Windows or Linux, and so on? When the legacy features are inside same project, and you replace old libraries with new ones, or add new OS support, you automatically allow legacy games take advantage of those libraries or run on new OS. With legacy features separated, you would have to update two projects differently, and something tells me no one will bother to update legacy one.

You are calling AGS code "aggravating" and hard to understand. I agree it often is, but that is not only, and even not so much because of legacy stuff, but rather simply because the code is bad. Besides actually outdated stuff, there is old stuff which is still being used (simply because there are no alternatives), but it is coded no better (if not worse) than outdated parts.
AGS always needed restructuring and refactoring. While I was still working on it I wanted someone to help me do it so much, but unfortunately no one did. And now, when there is an interest in this again, it appears one find it easier to just throw away the code which they do not understand. That's very sad.

(That's ironic you mentioned "coord multipliers that make your brain melt", because these caused lest trouble in comparison with everything else. After I found out what they do, I simply ignored them.)

Some of the legacy features you said you want to cut, are still used. There are long-standing members of AGS community who make games only in 8-bit mode, for instance, and these games are not some random crap, they are good games that earned AGS awards.

I believed back then, and I still do(!), that it is possible to solve "legacy" problem by restructuring the code. There are examples of how it was partially solved along the way. Take the conversions done during game loading, that you are removing in your branch. They are already picked out and grouped in distinct functions. They do not cause trouble! The conversion makes old format be loaded into new format, and the engine later works only with new data.

Similar restructuring could be done to all the legacy checks throughout the realtime functions. Out of the top of my head, engine could have distinct runtime classes to work with new versions and old versions of the game. For example, something like Character class and LegacyCharacter class.

Of course, that does not mean that those legacy features need to be kept in the Editor. Since Editor is mostly for making new games, that will be less painful to remove these unused options from it.
But the interesting consequence of engine's compatibility could be that one could use newer engine with older Editor.

One last thing to mention here. You said that you aim at making code easier for developing AGS-inspired engines. The thing here is that if one is making totally new engine, they do not need to see most of AGS code. Much of game objects and gameplay can be easily recreated just following basic concepts, they do not have to copy code from AGS. (Unless these engine writers are beginner coders). There are already multiple new engines written by people who formerly used AGS (Adore by Steven Poulton, Gobby by "Journey Down" team, Adventure Creator for Unity, MonoAGS currently being developed by Tzach Shabtay), and it does no look like they suffered from not being able to take particular code from AGS.
On the other hand, if someone would like to write new "AGS games player", to run all AGS games with, they would have to search through more than one branch to find how old game support was done. And since the code in old branch (the one containing legacy features) still will be unrefactored, that would be more difficult for them.

Take recent Scotch's announcement on forums, for example. He said he's writing a practical replacement for AGS engine on Js/WebGL, *and* he is aiming at supporting running all games, uncluding very old ones.



Of course, what you do is an interesting experiment, and may be very useful to improve your understanding of how AGS works. But I urge you to consider the alternate option I mentioned above.

With the route you chose there is of course a "third option": clean the code to make it easier, refactor, and then find way to get legacy stuff back without screwing existing things up.




But there are other considerations too.

If I try to gather worst problems of AGS now, including those that make updating it difficult, from *my own perspective* the list would be this (in no particular order):

1. Outdated back-end (Allegro 4 and accompanying libraries - font, sound). Makes it hard to support contemporary platforms. Needs to be replaced completely.
2. Outdated software renderer. In many cases supporting it requires writing software drawing methods (like alpha blending, for instance) duplicating hardware-accelerated functionality. Makes it nearly impossible to support advanced graphic effects, like fast scaling and rotating (including camera effects), and so on. Needs to be thrown out, and 8-bit support reimplemented in some other way (like shaders).
3. Generally bad state of code. Must be refactored.
4. Bad plugin interface. It is generally not safe, and also literally prevents from modifying certain engine's internal data structures, which is ridiculous and must not be tolerated. But replacing old plugin API is easier thing, because plugins are attachable part of game, and may even be recreated. For example, @jjsat successfully recreated few plugins to make them portable.
5. Savegame format. Not only it is badly structured, and writes redundant data, the order of data there makes it hard to change some things (like remove object limits). Savegames may be safely rewritten from scratch, because in AGS there is rarely a need to maintain saves compatibility between major releases.


If you look at this list, the things that you are planning to do barely solve anything, unless making refactoring bit easier. At the same time, all those big issues will remain.
Of course you may say, that after you remove lefacy stuff, it may be easier to work on those problems too, but I do not think these issues are related much. The engine consists roughly of two parts: the game logic and back-end (renderer, audio players, etc). Removing certain stuff from one part won't necessarily solve anything in regards to another.



That's basically what I wanted to say. Maybe I could add something else, but it seems too much text already. I tried to discuss all this on forums many times in the past years (I do not remember if you've been around then), and it's hard to repeat myself again.

Sorry if my ramblings cause more confusion or discourage you. I'd really wish if other collaborators, like @sonneveld, who was working on SDL replacement, say anything, but everyone seem to be silent so far, which makes me think that either they agree with your plan, or do not believe other way is worth trying.

**Comment by AlanDrake** [comment]
> AGS always needed restructuring and refactoring. While I was still working on it I wanted someone to help me do it so much, but unfortunately no one did. 

That is not surprising, I had wanted to do it in the beginning, but it was like wading through the jungle and  I had not enough experience with c++ for such a big project, so I got discouraged and just ended up hacking together the few features I needed and called it a day. After all I had more interesting and rewarding things to do.

> (That's ironic you mentioned "coord multipliers that make your brain melt", because these caused lest trouble in comparison with everything else. After I found out what they do, I simply ignored them.)

Ironically. that's what caused me most trouble so far, it was like that code was actively fighting me for survival. The whole concept was wrong, because it was designed for 320<-->640 in mind. It's also detrimental to new games, like for that poor soul on the forums that increased the width to 427x240 and suddenly everything was weird and broken. This is not acceptable.

> Indeed that's partially because you are going to throw away all the functionality to run old games, which we were gathering for years, piece by piece, spending so much time just to let people play them on contemporary platforms. 

And all that stuff will live on in the classic branch. I just don't believe any source developer should suffer to implement new features trying to keep compatibility with the age old.
AGS was born for letting people make new games, not for our inner circle to play older games. That's a plus.
New game developers are not attracted to the tool because it plays older games, but because it enables them to make their own.

> I feel like your decision is mostly dictated by the wish to create an "easy" version of en engine for your own purposes, without forethinking of any future goals.

Yes, I admit it, I do this for selfish reasons, I want AGS source to be more mantainable and easier to bend to my needs.  
And that's a perfectly valid reason, it's what motivates me.  Without that I'd simply have no incentive.
I can't refactor the current code, but maybe, with a stripped down AGS, I can make an attempt.


Anyway I'm still not entirely sure what to do about 8bit, I'm leaving that for last. I have fond memories of palette magics from Deluxe Paint days, but very few people can pull it off today (the only one developing a 8bit game I know of is Scavenger). It's also likely we'll come up with a better way to implement it, so we'll have to scrap away all the old code anyway. We'll just have to see.
Were it for me I would even destroy the text parser. God helpe me, I *hate* text parsers.

At the moment I'm in the process of removing the older script commands, that is probably going to keep me occupied for a few days.

Sonneveld said he was busy with RL stuff, so he had to put the SDL branch on hold, at least for the moment.

**Comment by ivan-mogilko** [comment]
Very well then.

If you wish I can add you to the repository contributors to let you push your commits here without waiting for someone else to check and press a button, because I likely will be busy with other things in the future.

**Comment by AlanDrake** [comment]
Yes, I wouldn't mind to be back again as a contributor.
Regardless of what happens with my cleanup attempts, I'll be at least able to merge hotfixes, add tags to issues and such.
I see everyone else is pretty much like me, active in short bursts, we should cover each other since there aren't many of us.
I'll try to avoid 3 years gaps this time 🐌

**Comment by ivan-mogilko** [comment]
@AlanDrake, did you receive the committer invitation?

**Comment by AlanDrake** [comment]
No, I didn't.

**Comment by ivan-mogilko** [comment]
Our team page sais there is a pending invitation to you... maybe you have a notification somewhere?

**Comment by AlanDrake** [comment]
Oh I saw it now, sorry I was expecting a notification or mail, perhaps having been a member before silenced it.

**Comment by AlanDrake** [comment]
So, I stumbled into interactions.h/cpp and realized there's a whole new layer of legacy complexity rooted deep down.
It doesn't look like I'll be able to just cut it away and be done with it, nope, I've been staring at the code for a couple of hours, and I'm still unsure how to divide and conquer this mess...

**Comment by ivan-mogilko** [comment]
If you are removing legacy things, interactions may be eliminated as class, because they were fully replaced with event handlers at some point (in 3.0 maybe? don't remember). If you look around, whenever an interaction is mentioned there should be respective event handler nearby. This pair goes in every class that has some events to script (rooms, objects, etc).

**Comment by AlanDrake** [comment]
Thank you, that helped. I actually managed to make the engine compile without breaking serialization/deserialization.
Now it's the Editor's turn...

**Comment by ivan-mogilko** [comment]
I would like to try restoring another one of my older branches, with new savegame format. A while ago I spared some time to filter out data that should not be in savegames (because it never changes) and regroup the rest to make it look comprehensible.
Did you have any particular plans for this that might be taken into account?

**Comment by AlanDrake** [comment]
No, so far I haven't any specific plan for the savegame format.
But I have a long term plan for serialization in which I'd like each component (and subcomponent) to have its own version and be able to upgrade itself.

>What do you refer to as "global vars"?

The Global variables from the editor's panel, I didn't even know/remember they were considered legacy until I read the comments.  And I even used them in my game back then, shame on me. Although, I have to it's comfortable not having to manually import them everywhere. But if it's deprecated I should get rid of it, I guess, which will stop other people from getting confused.

**Comment by ivan-mogilko** [comment]
> The Global variables from the editor's panel, I didn't even know/remember they were considered legacy until I read the comments.

How are they legacy? That's practically visual programming. Lots of people use these and consider useful, especially beginners. Did you consider asking on forums first?

Maybe you confuse the global variables pane with "graphical variables" in script API? They are different thing...

**Comment by AlanDrake** [comment]
Oh, I haven't stumbled in the part that converts to a hidden module, I guess I'll try to wipe all those OldInteractionVariable/InteractionVariable/globalvars[] and see what happens.

**Comment by rocifier** [comment]
I would be interested in helping with the cleanup but we would need to setup some strong guidelines and restrictions for it before others can contribute. What do you think about setting up some kind of roadmap or at least a TODO list?

**Comment by AlanDrake** [comment]
Sorry if I'm not being much alive at the moment, I have too many interests preventing me to focus on a single project, I'll surely get back on the piece eventually.
But in the meanwhile, if you want to prod me and talk shop, I suggest you join the discord channel. Talking about the stuff might fool me into putting some effort.

**Comment by rofl0r** [comment]
i fear that this project will not lead to the death of legacy, but to the death of AGS. as soon as ivan's no longer around, nobody will keep the "legacy" AGS alive, so it will not work anymore on the next incarnation of operating systems.
while the effort put into a legacy-free AGS will only benefit a handful new games, if even that.

I believe time/effort would be better spent by improving upon the existing codebase, while keeping it compatible. I'm sure there's tons of places where one could put energy into and still get reward out of it, without throwing everything that's not seen as chic&modern over board.

even if someone else will keep maintaining the legacy code, having such a split (as was seen in the past with python2/3 and lua 5.1/5.2) just divides the community in an unhealthy way.

**Comment by AlanDrake** [comment]
I see your point, but the death of AGS will only happen when no developers will be there to keep it up and running. A discouraging codebase prevents many from even trying. I've taken various things apart, and there's various stuff that's tricky and just doesn't works as you would expect it.

Adding new stuff to 3.4 is no simple task, and refactoring isn't a simple job. Whenever I tried I just burned out, so from my point of view a cleanup is vital.

Besides if we actually manage to clean up and refactor the codebase, it could help in the future to upgrade the legacy someway.

**Comment by ivan-mogilko** [comment]
The death of AGS will happen when no one is making games with it :-) (I believe saying this back in 2012 in one of them "future of AGS" discussions).

Anyway, I will probably repeat myself again, but my opinion only strengthened, that there is little benefit from clinging to old code. Even if you manage to clean it up and refactor somehow, all the main problems of it still remain: it's based on old technology (libraries), script with limited abilities which felt non-contemporary years ago, weird design choices that root deep into the system.

On the other hand, if you do not see keeping legacy support a priority, existing code holds even less value, because most of the basic stuff which makes a foundation of adventure game could be recreated without even referrring to AGS code. Then, some of the components may be plain copied over from AGS, like the script interpreter, if you wish to keep same script (it is practically a stand-alone program, connected to the engine by its API).

Personally, I've came to the conclusion that the engine needs full rewrite from scratch several years ago. I've even started to work on alternative, writing it on top of Allegro 5 library (but that's only one of the options). I had to cancel that work though, realizing that I will have to cease work on AGS itself, and people kept asking me to fix/improve things in it for their games (besides, at the time no one mentioned to be interested in helping me). Still wondering if that was a mistake on my part.
I've returned to this though earlier this year, and then learnt that several people are already writing their own alternatives to AGS, or have thoughts to start these. Personally, I'd rather join one of them if I had an incentive to continue doing anything AGS-related.

**Comment by jdiperla** [comment]
Well CW I think over the last few years you have been the heart and soul of AGS. I would rather see support for older games be separated from supporting a new revised engine. And I think the priority should go to the new engine and support on the legacy stuff should come only if there is time to do so. Can we do something like that? Can the developers put their foot down, hold their temptations and place and start something like this?

CW, what wold you write the new engine in if you had a choice? Would you stick with C++ and C# or would you move away from that? Most people nowadays focus on porting. Which means Mac, Windows, iOS and Android for a certainty. Would that mean moving from Allegro 4 to 5 or even to SDL 2? OR would you do a rewrite on a more web based approach such as phonegap(Javascript, HTML, CSS).

I think it's important to focus on keeping things familiar with the current AGS and what makes it great and for why people use it. Otherwise it will be a different engine all together.

**Comment by ivan-mogilko** [comment]
@jdiperla, I do not feel this is a right place to discuss the questions you mention. I posted here just to state my opinion about very particular idea (continuing developing AGS with the existing codebase). I would not want to "steal" this thread or make it deviate onto another topic.

**Comment by fdslk1** [comment]
Hello fellas, I wasn't going to comment on this, since I'm not developing for the engine itself, but I'm a user that actually use AGS and loves it.

If you want some input from a user that actually uses AGS, I'll just pinpoint some stuff that I've seen asked several times by other users that I think will come in handy in your decision making.

- A lot of users want and need better mobile support. Or at least better documentation. You can see an apk ready to download in some forums, but most users are unsure what to do with said apk or how to actually compile to mobile. So a native support for compiling to mobile (Android/iOS) in the binary version of AGS could come in handy.

- AGS is a hell of a software, that can handle a lot, but with mobile at current state, a lot of other users have asked if it can be ported to HTML5, which could even be a better solution than just going mobile. With the current state of WebAssembly, I'm not sure whether that's possible or not, but it will certainly attract a lot of new users and keep the user base at large.

- AGS may feel dated for a lot of users since it works mostly as old school game engines. The only thing I would add to it, and most people would find useful is scability to sprites and objects within the scene editor and parallax scrolling.

The rest of legacy options, I think CD Audio can be removed since that is becoming old tech nowadays. The 8 bit and palette system I would encourage not to remove since the very grace of AGS is the support  of retro pixel style, which as most know: something people actually buy.


Anyhow, this is what I learned from other users, google searches and forum scavenging and that I see in most engines now to be added. I hope my insight comes in handy to you.


AGS is a great engine, it won't die while it covers better porting to other mediums than just desktop.

**Comment by rocifier** [comment]
Hmm as a dev interesting in contributing this is a bit discouraging. A cross-platform rewrite would be an enormous undertaking and I'm not sure if there are enough interested devs floating around who know javascript. The thing that attracted me in the first place to AGS was its old school code base and its support for legit retro games on modern operating systems - not just "retro styled".

**Comment by ivan-mogilko** [comment]
> A cross-platform rewrite would be an enormous undertaking and I'm not sure if there are enough interested devs floating around who know javascript.

I heard there is already a javascript rewrite under way (cannot give any links though), and there is an alternate engine created in C# by one of the AGS users which seems to be already in usable state (incompatible with existing games though). I mean, I understand that writing an engine from scratch is a serious undertaking, but, after all, adventure games are simply animated sprites walking around screen, it does not take 10 years to write an engine for that.
But then again, it's all about priorities and goals, I guess.

> The thing that attracted me in the first place to AGS was its old school code base and its support for legit retro games on modern operating systems - not just "retro styled".

I am honestly confused by this, what is "old school code base"? The code in AGS is simply _bad_, which was very frankly admitted by its author. Similarily, I am also not sure what is the difference between "legit retro games" vs "retro styled"? With all honesty, I cannot think of any reason why one cannot make a retro game on modern codebase. Indeed there may be certain difficulties, like running 8-bit games on modern 3d engines (if that's what you mean?), but there are ways to solve that.

Also, as I keep noting, the current engine has more and more problems running on modern systems.

**Comment by rocifier** [comment]
By old school I just mean C++ mixed with C. I guess you are right about modern engines being able to produce that style of retro games, where you actually can't tell (fake low res with nearest-neighbour sampling etc.)

If I was going to do this I'd use pixi.js which has masking support for the novel walkable masking technique we use, and deploy it using https://crosswalk-project.org/ with that combo we could probably get 60fps on most devices.

**Comment by fdslk1** [comment]
> Anyway, the meaning of what I was saying earlier was that, in my opinion, it is too late to continue refactoring AGS code "in place". You will spend lots of time, but will have a lot of same compatibility and system support issues in the end.
> It may be better, and perhaps even faster to recreate it, using contemporary backend libraries, and more convenient program design. That, even if you want to just have same behavior in the end.
> 

Honestly, I think for cross-compile the only thing AGS needs is the ability to port to HTML5. That's my mho

**Comment by rofl0r** [comment]
html5/javascript are quite bad technology, even though it is currently hyped...
but in order to use it as a platform one needs a bloated, memory-hungry browser.
example: firefox sources ship in a 200 MB compressed tarball, it takes an hour to compile on x86 system, and more than 4GB RAM. at runtime, the browser will need at least 200 MB, plus a lot more if you want **fast** javascript, which means there needs to be a JIT. that's only the baseline to be able to use JS at all. the game itself may then consume lots more. not exactly what you would want to run on some slim gaming console attached to your TV, etc. if you want **fast** code that's cross-platform, your best option is still C/C++.
and that's exactly why still most of the more complex games are written in those languages.
note though that even C++ will use a lot more RAM than C due to its coding style which heavily relies on dynamic allocation of new objects.

**Comment by ivan-mogilko** [comment]
I found javascript a nice and easy language, but I only learned it recently, and have no idea of how well it performs (or how HTML5 performs). Since @jdiperla asked above, I would actually choose C++ if I were doing rewrite, but then again I may be biased here because I am very used to this language.
There is already [an "alternate" adventure engine written in C# (Mono)](https://github.com/tzachshabtay/MonoAGS), but I did not check its system requirements yet.
It's hard for me to tell why, but many people seem to like the idea of playing games in their browsers, and that is where js/HTML5 is required (or perhaps it's simply seen as a simplier way to get it crossplatform?). Maybe there is a sense for having a js port alongside with "main" C/C++ program, but then it has to be supported by another group of developers, because maintaining both at the same time would be overkill.

**Comment by jdiperla** [comment]
I think the export to HTML does mostly have to do with the fact that the game would then be available on multiple platforms. I have heard many folks on several forums say that Javascript performs significantly faster than C++ as well. It also seems to be that a few of the modern systems are making less and less use of C++ and I suppose it is feared that it may disappear almost entirely within the next decade. This is of course debatable. I think with Cordova/Phonegap/NodeJS... many feel that creating something in HTML5, CSS and Javascript will allow them to reach all the necessary platforms for their apps: Windows, Linux, Mac, Android, iOS and Universal Windows App(X-Box, windows phone, etc..). 

Honestly, I spent the $70 for adventure creator on Unity because of it's cross platform capabilities. I have been creating an adventure game that has taken significantly long to do in AC. I am honestly torn on what to use as an engine, so I decided, what the hell? Let me try and make my own in Javascript. So I began experimenting in Javascript with Phaser and Easystar. Using software like ExtJS and Codemirror, etc... You can even make a cross platform editor that will work on desktop and mobile too.

The downside of using web technologies however is that there is no easy way to compile to a console or handheld video game system as far as I know. I think it would be possible to write up a plugin for Unity maybe that can import the game and then compile it to those consoles since Unity can use Javascript. But that may need to be a side project for somebody to do. The other issue is that web technologies leave your source and data  open and exposed. Some game Dev companies may not like that idea. I know code can be obfuscated, but I believe that it can easily be reversed. One site offered a solution of obfuscating code by hiding in image files and that it was impossible to de-obfuscate that way. Not sure how accurate that is, but that is what they claim.

But going back to being torn about the engine to use:

AC: Powerful, some what user friendly. But unless you know Unity very well, you will only make a basic type adventure game. Unity is free, but has different license types that make it confusing and limiting.
Visionaire: The closest thing to AGS, but the licensing and purchase options are astronomically stupid.
AGS: My preferred engine of choice, but porting to anything else is just not easy.

I feel what is holding back AGS right now is also the fact that we are focusing on Legacy support and thinking of refactoring the code. I think it has done a decent job of going way back on legacy. And CW and others have dedicated and done so much to this engine. But I understand CW too when he mentions he wants to give up on upgrading AGS further. It just seems to be stuck in 2004. But maybe that is something that SCUMMVM should handle going forward. I think, whether C++ or C# is used or even Web Technologies, whoever develops AGS further should re-invent AGS to something a little new behind the scenes. It should have that familiar AGS feeling and ease of creating adventure games of course(We need to know, this is still AGS), but updated to have fewer bugs, more features, updated technology and cross platform capabilities. Maybe instead of going with the usual versioning system of 1,2,3,4... We can go with the year instead. For example: AGS 2018, the next versions: AGS 2018.1, 2018.23, etc... 

I don't know. I love AGS. Nothing about the engine except it's lack of easily porting the games bothers me about it. If we can just solve that problem, I would be happy for a few years ;-)

**Comment by ivan-mogilko** [comment]
> I have heard many folks on several forums say that Javascript performs significantly faster than C++ as well

I wonder if they realize that's impossible for interpreted language to perform faster than the language it is interpreted on? :P

**Comment by jdiperla** [comment]
LOL, good logical point. This is why we need you CW...

I guess it is also how well you write your code. Bad code in C++ can run slower than great code written in Javascript. I believe Game Maker Studio uses Javascript. As far I see, even if it is not faster than C++, it runs very effectively on most systems. The same with other games written in Javascript. So I don't see performance being the most significant issue of moving to it. For example, see this engine demo written in Javascript that is a few years old: http://jsgam.sourceforge.net/spookyadv/

Link to source: http://jsgam.sourceforge.net/

I believe it is more the code being publicly viewable by anyone and also porting to other systems like consoles.

I have found encloseJS that seems to compile to machine code. And then I also found that Emscripten(LLVM to Javascript) can convert C++ Projects to Javascript. Of course, as with anything, who knows how effective it is and how easy it is to use.

**Comment by fdslk1** [comment]
> I have found encloseJS that seems to compile to machine code. And then I also found that Emscripten(LLVM to Javascript) can convert C++ Projects to Javascript. Of course, as with anything, who knows how effective it is and how easy it is to use.


I tried to use Emscripten in personal projects is kind of a mess, but I'm not a pro coder so I'm not sure how good I am. Still a lot of game engines are using it to port to HTML5. Unity and UE4 use it.

The good thing about the builds made with Emscripten are how optimized they work. If anyone is willing to give it a shot or check it out. I think the main issue of people wanting to port and distribute their games with AGS will be solved...

Lets be honest folks, you need a lot more than just windows and linux to actually make a profitable game in the indie market.

If AGS gets a better way to port the game games, that will be the game changer. Or, if you want to avoid doing that, someone that actually knows how to do it, can write a proper documentation or tutorial.

**Comment by rocifier** [comment]
Performance with JS isn't an issue as long as you have access to WebGL and V8. Then it would be overpowered for any AGS game. 200MB memory is nothing considering most phones coming out now have 4GB RAM built in. I've not seen any on the market which have below 2GB, I went and checked the other day. No point in designing a new engine for legacy devices.

**Comment by SuperDre** [comment]
Except you really underestimate the ammount of phones that are out there which still haven't got 4GB ram (that may be common with new phones, but certainly not phones/tablets which are older than 6 months). These days a lot of people just buy their phone and use it for several years. Also don't underestimate the ammount of computers/laptops that are still in use.

Maybe you buy a new phone/tablet/computer every year, but most people don't.

To be honest, I don't see the need for another new engine when you've already got enough other engines that are excellent at also doing adventure games (both 2D as 3D).

So it's rather 'no point in designing a new engine' period.

Better spend time at developing a better designer which works with already available engines.


Also JS isn't really a great language, yes it's much used, but that doesn't make it good.


________________________________

**Comment by AlanDrake** [comment]
This is going off topic.
The current AGS is feature complete and has a working IDE. Making another engine or a full rewrite doesn't seem feasible at the moment (we don't have enough experienced developers, and it should probably be talked about somewhere else).
Besides there are already tzachs' MonoAGS C# and scotch's JS AGS spinoffs that could use a helping hand.

The question is what do we do with the current codebase ?
If it needs to be mantained I still think there's value in cleaning it up.
If we stripped the code to the bone, it would be possible to finally organize everything into proper namespaces. From there we might rewrite components and refactor.

I don't know when I'm actively resuming my work on this branch, my coding sprees come and go, but if 1-2 people join in the effort we can slowly grind forward sharing our knowledge and ideas.

We can discuss and coordinate in the AGS discord.

**Comment by tzachshabtay** [comment]
@SuperDre I think you underestimate the amount of time a rewrite is going to take. I'm working on my c# rewrite for over 2 years now, I think (https://github.com/tzachshabtay/MonoAGS), and realistically, if nobody joins me, I won't have a stable release in another 2 years.  Also XAGE (http://clarvalon.blogspot.co.il/) has been in development from at least 2008(!), and still not released. By the time a new engine is released and actually used, low powered devices will probably be a non-issue.

I feel Javascript (especially if it's typed- i.e either TypeScript or Flow) is a very good choice for a rewrite, not because it's a great tech (it's really not, and LOL about it being faster than c++) but because of its very wide cross platform support, and up until recently the only reasonable choice for web.
Like CW briefly mentioned, Scotch is already working on a javascript rewrite, and iirc he said he had more than 70% of the APIs already implemented. He also said he'll publish his work on Github at some point, though so far I can't see it. I'll try poking him on Discord.

**Comment by ivan-mogilko** [comment]
@AlanDrake
> The question is what do we do with the current codebase ?
If it needs to be mantained I still think there's value in cleaning it up.
If we stripped the code to the bone, it would be possible to finally organize everything into proper namespaces. From there we might rewrite components and refactor.

@tzachshabtay 
>  I agree that there's value in cleanup, though would probably be more effective to do the cleanup gradually instead of removing everything at once, which kind of leaves the current AGS codebase in limbo.

Tzach, honestly, I've been following the advice you gave me for few years, about refactoring in "baby steps", and it caused me lots and lots of pain. 

I wanted to mention this before, but probably now it is better time. In my experience I found there are two refactoring cases. First when the code is generally okay but some parts are bad, and you want them better (this is often the case during development of a new code). In that case doing "baby steps" works well. It also works well when you are just getting acquianted with the code, so you just refactor a bit here and a bit there, and that works fine.

But there is a second case, when the code is so bad and entagled so much, that it takes additional effort to keep it working after refactoring even a small bit. This is what happened to me somewhere during refactoring of AGS, and also one of the reasons why I ceased doing big refactoring in the past: I had to spend more time trying to invent the way new refactored code may seamlessly work with the parts that were still unrefactored. I also had to artificially restrict the changes I made, and spend extra time and effort designing "intermediate" states of code (Not to mention I had to redo numerous tests after each such step).

Indeed, doing gradual refactoring will also keep codebase in working state at most times, but this will also increase the total time of work, and risk of fast burnout. Meanwhile, the number of useful changes to the engine will be kept at minimum.

This is why I'd rather propose different approach (used it alot in another project):
1) First, come with the final design, on paper. What parts would you like to see in it? How will these parts connect with each other.
2) Locate where the analogues of your components are located in AGS, just to know where they are and how they are interconnected.
3) Code one full component of the final design, even though as a first prototype.
4) Cut out old component from program; not necessarily deleting it, e.g. just replace function calls with calls to your new component.
5) Make sure new component works new way, then, if wanted, bring back old behavior, with perhaps a "legacy mode" switch (there are different ways to do this codewise).
6) Repeat with next component.

**Comment by tzachshabtay** [comment]
All fair points, but I don't think it contradicts my suggestion for doing it gradually, i.e, like you said, one component after another (it's still "baby steps", just the baby is maybe a little bigger). From what I understand, @AlanDrake simply went "Leroy Jenkins!!" and started cutting everything (i.e removed stuff horizontally across the board), which might be fine if there's a quick test & merge back, but if work has halted and nobody else is helping and people might want to merge other stuff into master, it will just get more and more difficult to merge it in. It might be worth it to focus on this and freeze other development until something can be merged back.

**Comment by AlanDrake** [comment]
Yeah, sorry I lost steam there, but the cutting was proceeding quite well, if someone wants to join in I'll free up some time to merge the latest master changes, make a list of what has been done and still needs to, and move the 4.0 branch to the main repository, so it's more visible.

Just please stop changing the game format, I thought we had agreed that 3.4 only needed band aids to keep it working.

**Comment by ivan-mogilko** [comment]
> Just please stop changing the game format, I thought we had agreed that 3.4 only needed band aids to keep it working.

Really, everytime someone mentions "we" here I keep wondering who is included into this "we".

I can't find a way to put this correctly, but this baffles me so much, that I work for several years almost alone here, and then people start coming and telling "now we will be doing this and that", and apparently all I can do is submit.

Is this a development plan? If so, why is it not posted in a more elaborate way? Who decides what to do now? Whom should I listen to and what guidelines should I follow if I need to make a change? What if someone asks me to add another thing into engine, should I tell them I cannot do that because Alan Drake told me not to? Or perhaps I should just do that in my personal repository and post as a custom engine version.
I am simply loosing understanding of what is going on.

**Comment by AlanDrake** [comment]
Sorry, I didn't mean to anger you. My memory isn't worth much, but when you said you were planning to wrap up the RC version and be done with it, I'm pretty sure a feature freeze was discussed. 
Altough, yes, I cannot say that anyone expressly agreed to anything, except that you'd keep a 3.4 branch all for you.

Since I've butchered various places that handle the game project data, but I'd like to keep compatibility with your final release, the less the game format changes the better. I mean, merging is going to kill me, and I know... it's a grave I've dug myself!

Slack was the place to discuss internally for this stuff, but when you left, it kinda died down.
I'm hoping to bring wannabe devs on discord, since it's one of the few things beside slack that work under this evil corporate proxy and I can use to be present.

>What if someone asks me to add another thing into engine, should I tell them I cannot do that because Alan Drake told me not to?

I'm not sure how to answer this, since you said yourself you were allegedly finishing the release candidate before abandoning the project (a day we all dread of), what are your actual plans ?

**Comment by ivan-mogilko** [comment]
Sorry, but I decided to delete my previous comment. I do not feel belonging to this project anymore, and don't have any actual plans.

**Comment by jdiperla** [comment]
OK, that's it CW. Look, you have done tremendous work. Tremendous. But I can not condone these tantrums you display from time to time. I can not sit idly by while you speak that way of the rest of the community(Even just one person). It is uncalled for and disrespectful. I have been apart of the community since 1999. I log in nearly daily and I have seen Chris as he developed AGS right from AC Creator. Never had I ever seen him disrespect a member of the community. Never had he ever displayed himself in any manner that lacked humility. And let me tell you, he had plenty of reasons and right to lose it. But never did. He always thought of the community and did what he could to make everyone happy.

Now I am not saying that every developer needs to be that way. But CW, no offense, these tantrums and these "I am done" posts are getting old. Alan Drake has done some great things with his version of AGS. He offers great input and wise suggestions. You on the other hand ask for input, but do not want to hear it. People offer it and you do not want to hear it. Enough is enough. You have these moments of feeling so superior to everyone else where AGS would be nothing without you. You make these statements about leaving the project as if you are trying to threaten the community. I will not deny that the last few years you have been the heart and soul of AGS. I will not deny that we all appreciate what you have done and we all doubt at times that without you and your support, AGS would have been dead years ago. But that does not mean you have a right to these performances that you give. 

AGS first and foremost is Chris Jones product. If he were to come back and say: I want ownership back - then he he deserves it. But he gave up his years of work and gave it to the community as a whole. We should all be working together on this, whether we are developers, game creators or players of the created works. That means we need to listen and not be so stubborn. "You can have your say, but you can't always have your way". But you never want to hear what everyone has to say. You just want to hear everyone else say: "You are right, do it that way and we will support you. Also, you are awesome!" I'm sorry, but your "We" comment that you deleted really irked me. I can't believe you would say something like that about Alan or anyone else. So many people assisted in the continued development. Gurok, Monkey, JJS and so many others who are listed in the credits. But with your last comment, you just seem so self-centered that you probably do not even believe they deserve the credit. 

Granted I know all I have ever contributed to AGS is a large Simpsons Template, translation for another game, beta testing on a couple and part story writer for another. Not a lot. But AGS is still my engine of choice if I had to go with an engine. The only reason I am going with AC on my current project is because AGS lacks porting it's projects easily to other systems. But otherwise, I love AGS and I am tired of the same old arguments time and again. CW, you do not have the same love for AGS as some of us who have been here nearly 20 years have. You do not have the appreciation for it. 


You have a choice now. You can either Grow up, realize this is a team effort(Yes, maybe you do not have a lot of "developer" support, but you should still listen), stop whining and threatening to leave the project and continue doing some more good here. Or you can just leave and the rest of us will see what happens with AGS. Honestly, AGS has everything we need except the porting features. For the developers and players, we do not see what is behind the scenes or how the source code is tangled or whatever. We focus on the experience that AGS Gives us. 

Sorry, there is a lot of rambling here. But I am just so tired of his constant threats of leaving. Upset about his "WE" comments when there has been so much work put into this by so many others. Upset at how he treated Alan.. It's just enough. He doesn't want anything to do with this project... Then leave! Don't let the door hit you on your way out.

**Comment by tzachshabtay** [comment]
@jdiperla I feel your comment is really unfair.

Saying that CW doesn't listen? Unfair. Every issue posted on the forums gets a very fast response from him, including trying to reproduce, getting personal projects and working with them to fix the issues. He listens and tries to reason about each suggestion, and there have been too many to count. Even when people make silly requests, he still bothers to respond and explain.

Comparing to CJ: Unfair- not everybody can be CJ, not to mention CJ was working on his baby, and CW was working with somebody else's code, a much less rewarding job (and much more difficult).

The "We" comment- take a look in Github -> Insights -> Graphs and see his contributions: yes, there have been other contributors, but this is mostly a one-man show (especially after JJS left in 2014). And the contributions only tell half the story, the amount of man-hours invested by CW in AGS and the community is unthinkable.

Saying he doesn't love AGS like other people: completely unfair, if anything, he cares too much, which is what got him to this position in the first place. 

His comment might have been heated, and maybe he made heated comments in the past as well, but this comes from a place of real hardship and we should show support.

**Comment by ivan-mogilko** [comment]
I am really sorry about posting that comment, and I will try my best to not post something like this again.

But one thing I want to clarify, that it was never in my intentions to threaten anyone or to show superiority, although I have to admit that they may have looked like that from some perspective.

The only emotion that causes me to write things like that is a despair of realization that no matter how much time I spent there I could not reach the goal, and this irrational jealousy of not being able to continue work along other people. Because when we just started, it was my hope to be a part of a long living team of contributors, but since that team dissipated, I could not help but feel the grief of being left alone. And now, when I lost all hope and enthusiasm for this project, I should've just wish the good luck to however continue this job, but instead I have this irrational fear that I could never be a part of something that big and interesting again, and it makes me return here and make these stupid statements.

**Comment by jdiperla** [comment]
@tzachs Look, I apologize for my comment. It was not intended to be unfair
or cruel. It was heated and it was wrong. But it also had some legitimacy
to it. I agree, he is happy to oblige anyone that reports a bug or makes a
feature suggestion no matter how redundant or stupid it is. AGS is still a
strong contender in the AG scene because of him. Over the years I have
always acknowledged his hard work, even giving him a personal award on the
AGS Blog in recognizing him. I should not have compared his character to
that of CJ. I didn't think of it that way. But you are right.


Where I am struggling is with his comments about leaving and where he
criticizes others that are trying to help in whatever way they can with the
development. I can't even begin to imagine how hard it must be to take
apart and put back together the source code for AGS. Kudo's to him for
doing that. But you are right, it adds to his frustration. If I had that
kind of task ahead of me, I am sure I would feel like quitting too. But the
problem is still CW here. Even though he will oblige a feature suggestion
or bug fix, he struggles to take advice on the development of AGS. He has
standards that are different from others and he is not willing to work with
any other way than his own. That's fine to a degree, but because of that we
have become completely dependent on him for the development of AGS. I guess
this is fine. But then don't make posts saying "I am going to leave, I
don't have support, etc.." because he might have that support if he was
willing to adjust his approach on the development of AGS to that of the
other developers. It also discourages others from using AGS because they
may think: What's the point? CW will end development and noone will pick up
the torch since noone helped him in the first place.


I do still resent the "WE" comment as it did sound like a blatant disregard
to other previous developers. Many worked very hard to continue developing
AGS and many want to do what they can to help. And if I am not mistaken,
much of what has been recently implemented into AGS were features that were
in the Draconian Edition. Maybe the commits do not show that. Gurok was a
big help too. Monkey is trying to get things going on the Android side of
things. Wadjet eyes offered insight on a lot as well. Maybe CW did the
most, and is the only recent active developer on this, but he can't take
claim to there not being a "WE". And if he let up a little with his
criticism of other developers trying to help him, perhaps he would have
more help from others.


Again... I very much appreciate what he has done. I even think he is a good
person and in reality I don't want to see him go at all. But it is
emotionally draining to hear: I Am leaving, now I am staying, now I am
leaving. And he does address this in his new post.


@CW: Seriously dude, do you not see the amazing things you have done with
AGS???? Are seriously doubting yourself???? At this point, you might have
done more for AGS than CJ has. AGS only stands a chance because of YOU.
Whether you believe it or not, AGS and it's community will be one of those
things that will go down in Internet History and Adventure Game Engine(And
Game engines in general) history. AGS is a pioneer in internet communities
and game engines. And you had a major part of that. In a sense, you are
immortalized. If there was a hall of fame for adventure games specifically,
you would be in it alongside Chris and many others here. If there is a HOF
and you are not in it, it is surprising.


I highly doubt that if you left AGS that you would not become a part of
something bigger. You just need to have confidence in yourself. That is
where you will fail. That is where I personally fail. But you are a hundred
times better than I am in things like this. But you have to learn to share
the playground and play with other rules too and consider others opinions.
I think that is where you struggle. I think if you posted things like: "I
need a break for a few months to focus on X, Y and Z." It's not a big deal.
We get it. You have a life. You have more important things to do. AGS is
not the end all of things. It is seriously a hobby. But you can not toy
with people's emotions in the sense that you say repeatedly "I don't want
to do this anymore, so I am leaving for good... after I finish this.." and
then disagree with others who want to continue the work and their methods.
It's confusing and it can be a little demoralizing and will stop others
from continueing the work because they are confused. There is nothing wrong
with taking time off or really even just leaving for good. But you need to
work it out so that others can still feel they are allowed to continue the
work or help.


CW, I don't know how often we have to tell you this, but you do seriously
rock. I personally don't want to see you go. But at the same time I don't
want to hold you back from other things either. We want you to stay,
really. We just want you to share the playing field is all. If you permit
me CW, I want to PM you and share some other thoughts that I think might be
of benefit to you.

-------------
Giuseppe DiPerla
Deep River, CT Congregation
Phone: 860-518-1500
Email: JDiPerla@GMail.com ; Giuseppe@DiPerla.com
Google Voice: 860-615-9428

**Comment by ivan-mogilko** [comment]
> I do still resent the "WE" comment as it did sound like a blatant disregard
to other previous developers.

Okay, I hate to continue this talk, because I realize that snapped again, and just already sick of myself. But since you are so touched by this... What I meant was completely opposite to what you think it meant.
I said that because I often feel like being NOT INCLUDED into "WE"; that there are some plans being made and new ideas enforced by other people, which I am not aware about.

**Comment by fdslk1** [comment]
@ivan-mogilko @jdiperla 

I know I barely made two comments on this stuff and I know I'm no one to speak, but I think you're taking stuff far too personal and you should both stop typing for now and chill the heck out.

On-topic:

Now: Quick question, have you all considered making a poll about the features you want out and what people want in the engine inside the forums? I think you should ask your user base what they expect and need from the engine, most will say they need better porting (which is the only downside of AGS) others maybe remove legacy and others maybe want nothing new.

Sometimes the best course of action is doing nothing and just gather intel. And this is coming from a long-term AGS user.

**Comment by AlanDrake** [comment]
I already got the input that matters from chat and discord, I have even taken out 8bit mode from the kill list for the time being. Except few things all should mostly work as it does now, until the cleanup is complete. What comes next should be decided in the future, when and if this part is done.

There's no value in making plans for the users now, it's the developers the current concern. There's not a single one happy of working on the code. Any cool stuff has been postponed, be it because the code kept changing making a merge difficult or that side effects kept appearing out of nowhere, this has discouraged me and many others.

Cleaning the code from the old stuff is the only path left. Coding for AGS must become easier or else devs  who can't invest much time will keep running away.

**Comment by ivan-mogilko** [comment]
Sorry, but could I delete the unrelated comments posted above from this thread? Since that's me who brought them here...

Only shortly speaking on topic:
1) The most critical issues users were noting over last couple of years, that I remember:
* ports (already mentioned)
* unicode text support (believed to be not a big issue if you make game on English, but as soon as it is another language, esp. non-latin one, one has to rely on hacks)
* various annoying engine restrictions: e.g. max number and size of sprites.

2) In regards to how cleaning the code relates to ports: that really depends on what has to be done in order to make the port. I think this could deserve a separate discussion. For example, @sonneveld 's work on porting AGS to SDL2 (which supposedly could make it run better on some systems) was mostly done by writing an intermediate layer "under" the engine, where he substituted Allegro4 functions with calls to SDL2, meaning that the engine code (dirty or not) was largely untouched.
If the porting is done by completely rewriting code, like with the javascript port, cleaning the original code may not be required at all, except maybe for studying reasons.

I would also like to point to an unfinished rewrite made by ScummVM member made 5+ years ago. It is not 100% compatible, but has much cleaner codebase: https://github.com/fuzzie/scummvm/tree/ags/engines/ags


PS. Sorry again for the drama...

**Comment by SuperDre** [comment]
In regards to ports, I would suggest to pickup the rewrite for ScummVM as that is THE most popular adventuregame hub available and it already has many ports.


And no Ivan, you don't have to appologize for the drama, you did one heck of a great job, and that's not easy.

________________________________

**Comment by rofl0r** [comment]
as far as @fuzzie's scummvm work is concerned, it is already very usable. one can play most games with it. the major issues it has is that 1) save/restore is not implemented (which makes it a bit hard to debug issues that appear further inside a game...), and 2) the pathfinder algorithm is buggy (it probably needs backports of the fixes that recently went into AGS itself). i contributed to her port in the past, but couldnt get the pathfinding issues fixed backed then - that's basically what drove me away.

btw, if someone really does a rewrite of the engine, i think it would be wise to consult experienced team members about new features, so mistakes made in the past won't be repeated - example: the plugin mode for windows DLLs. (shudder)

**Comment by jdiperla** [comment]
Disclaimer: Just for the record before everyone thinks I am a complete Jerk
here... We worked this out behind the scenes and we both apologized for our
behavior. At least I think we are good anyway.


Back on track to the legacy stuff... Out of curiousity, how far backwards
compatible is the engine able to run older games? And how far back can it
import game projects from AGS? Just curious about that.

**Comment by ivan-mogilko** [comment]
> Out of curiousity, how far backwards
compatible is the engine able to run older games? And how far back can it
import game projects from AGS?

Engine: loads games from AGS 2.5 and higher, but 100% compatibility is not guaranteed (there were no bug reports for a while though).
Editor: v3.4.1 in theory can load everything since AGS 2.5 too, earlier versions could only load 3.* projects.

**Comment by jdiperla** [comment]
So if development of going back further was dropped and attention was given
to other matters, would it be a big deal to the community? I mean, if
memory serves, AGS below 2.5 compiled everything in DOS anyway. It could
easily be played on any system at that point using Dosbox on any OS
actually, including mobile and some consoles. I mean, if work still
continued on supporting older engines, great I suppose. But maybe it
shouldn't be a priority? As far as importing projects from 2.5... That is
good work! But the same thing goes for that... Not many projects will be
around from earlier editions and projects can probably be imported into 2.5
and then imported into 3.4, right? Just a thought. This was probably
discussed, but there were 70 pages to sift through and I have read every
post, but don't remember this being specifically discussed at length.

**Comment by SuperDre** [comment]
Well I don't know about the current regular engine, but it would ofcourse be great if it just supported everything.

But I guess that's more the job of ScummVM.. Having to deal with dosbox has always been an annoyance.


________________________________

**Comment by jdiperla** [comment]
Don't get me wrong, I would love that too. Especially if I can get Rob
Blanc back to working natively and even on my mobile devices. But with a
work around possible, maybe it would ease pressure and give more time to
the devs to work on other needed features or fixes to the engine. I mean
it's up to them. But just thinking out loud about that. I am not saying
that it's off the table completely, but maybe something that is more of a
side project or less of a priority.

**Comment by AlanDrake** [comment]
Okay then, I finished merging master to the AGS4 branch. Later I'll write down what is the current situation and move the branch to the official repository.

I think the only big gotchas so far are that a project must be first upgraded through classic 3.4.1 and that the blank room template is broken (due to its extreme vintage format), I'll replace it with another until AGS can generate it on the spot.

**Comment by morganwillcock** [comment]
In terms of getting off Allegro 4, are you thinking to replace existing library functions with something back-end agnostic or are the way that game events are handled too specific to Allegro 4? I think you'd probably pick up more developers if the whole thing was stripped back to a new event loop and stub functions to support the existing scripting language (would also give the opportunity to evaluate the language at that point). Personally I've only dabbled in C++, but would be willing to contribute if the bar to entry was a bit lower and there was a obvious place/person to go to when guidance was needed.

**Comment by ivan-mogilko** [comment]
> In terms of getting off Allegro 4, are you thinking to replace existing library functions with something back-end agnostic or are the way that game events are handled too specific to Allegro 4?

Coincidently, I sent a message to AlanDrake only today.
Recently I was considering trying to approach this problem from another angle: write a simple rendering engine based on Allegro 5, and then copy only game-related logic from AGS on top of it. Naturally, even in case of success there will be compatibility issues for a while, most notably it will take some effort to bring 8-bit games to work.
But I believe that may be easier, or at least cleaner, than trying to shove another library into current code. Especially considering there is an attempt to clean and restructure it under way.

That's just a thought, so I cannot give any promises. If that won't work, I guess that will at least be a practice programming with Allegro 5.

PS. Why Allegro 5? No serious reason really, A5 and SDL2 capabilities look similar at the first glance (maybe A5 is a little more ambitious with number of addons); Allegro 5 has floating-point API for drawing which perhaps gives a potential to support smoother movements in the game world (people were suggesting subpixel rendering years ago when all this AGS upgrade talk started).

**Comment by morganwillcock** [comment]
I've always like the idea that the core engine be basic with optional features on top, so games declare any optional features they need. This also means it someone wants to reimplement the basic core in some other language, they have a spec to work to and games that don't use extra features are then compatible. So starting from basic engine does sound appealing for someone new, plus if any major changes like independence of engine speed and game speed need to be made, I can imagine this would be problematic to implement as is. I do have a couple of work related projects I'm working on for the moment, but if this approach has interest then I would like to assist once these are finished. Perhaps just a design document in the interim, but in terms of backend library, it seems most options are event driven and so equally different to Allegro 4.

**Comment by AlanDrake** [comment]
Ok, I couldn't do a clean rebase, but the "ags4" branch has been updated to master.
I fixed some whitespaces and reverted one of those useless variable name changes you mentioned.
I tested it against my two 3.4.1 test projects and they work as expected, so there good hopes nothing went amiss.

@ivan-mogilko It's time to decide if it's good enough to merge into master for future development.
Let me know if you have any doubts. 

The current issues are mostly down to data format, as mentioned, a project has first to be upgraded with 3.4.1+ to be opened.
Also the current game templates use outdated formats, so they should be upgraded (scripts too).
I bundled an upgraded blank room, although I think it's something that should be generated on the spot.

**Comment by sonneveld** [comment]
Hi guys, I know I'm late to the party, but I'm in the slow and gradual refactor boat. I don't think big rewrites or removing large caches of code is a good idea. It would take too long for a rewrite to be on feature parity, and removing lots of code would require a long time to iron out bugs.

I haven't had a look at this AGS4 branch, but if there has been some serious changes done, I suggest keeping it as an ags-ng branch and migrating changes slowly over the master.

I think writing some test game data would be helpful here.  Just to make it easier to test the behaviour of features (Current and new)

My SDL2 port is working quite well at the moment. That might be a candidate for AGS 3.5.

**Comment by AlanDrake** [comment]
If anything, removing old chunk of unused and backward compat code is only going to make it easier to iron out bugs.

Besides we're not losing backwards compatibility, it's going to remain on it's own branch.
AGS main purpose is to make games, so it's not a bad thing to make a clear cut once a while. Implementing new features while not breaking old behaviours is harder than it should be and discourages further development.

It's also pretty confusing from the user side having all those options and old script commands, how does a newcomer know what's the proper way to code for the last version and what's deprecated ? A newbie might be using a game template with old script APIs, then he goes on the forums and sees different things, confusing options, etc.
Experienced AGS users may know every nook and cranny, but we also want new people to join in.

Some game developers might also want to customize the engine to suit particular needs, so it would be nice not to make it harder than it should be.

Have you already tried merging your sdl2-port and ags4 ? What kind of impact would it have ?
I hope it wouldn't make it too hard since it's something I'd like to see coming to AGS.

**Comment by sonneveld** [comment]
quick thoughts:

Leaving legacy code in its own branch will mean it will wither and die.  The only alternative to run old games on new hardware is to hope that scummvm supports it.

In regards for ease of use for users, I'm not sure that's an argument for removing code.  That's an argument for marking functions as deprecated and allowing running in a forward compatible mode that doesn't expose deprecated functions.

Have you considering modularising the code?  Take old legacy code and wrap it in a module, allow switching modules to enable newer behaviour.  It could even be a compile time option for developers who want a custom engine.

I haven't tried merging sdl2-port into ags4 branch but I would prefer to merge it into 3.4 to allow a 3.5 release first.

**Comment by ivan-mogilko** [comment]
I wish this cleanup happened years ago when I still had enough enthusiasm to work a lot, and there were more contributors around.
Today I have a constant feeling that this project is simply dead, because no one has interest in actually working on the engine itself.

Both paths mentioned here have procs and cons, and I would be glad to give my support to either.... but the problem is, defeating the cons in both ways is possible only when there is a dedicated and active team.

The more I think about it, there is no a good way out of this situation. No matter what path is chosen, whether modularizing keeping the old code, or cleaning up, possibly breaking program, then fixing it and trying to get more updates... Who is going to do all the work in any case? Seems like no matter what it will lead to a program that does not evolve and has no proper maintenance.

**Comment by sonneveld** [comment]
Module development could be merged in piece by piece, all modules wouldn't necessarily need to be complete before being merged into default.

That doesn't mean we can't replace or remove some code.  Like the record/playback, it never was finished so we could remove that completely.

**Comment by ivan-mogilko** [comment]
Sorry for the above negative comment, but I could not resist saying that, because that's really how I feel right now.

After posting that I realized that probably should post something more positive too, so here it is. Sorry for the long comment below, but please read it all...

***

@AlanDrake , you are saying that it is hard implementing new features because of legacy code, and while I agree in general sense, it does not mean **that it is exactly the code that you want to remove** that makes it hard. It may be because the code in AGS is generally bad. It is also may be because of other things in the engines that cannot be removed right away. And removal of the legacy code won't necessarily make it easier.
May you give an example of a feature that you wanted, and tried to, but could not implement because of some of the legacy code code?

***

> I'm not invested enough to be able to modularize legacy from current, moreover an attempt like that involving a lot of code being moved around, would make it difficult to merge again if there's further development on master in the meantime.

This is a problem as with any change or refactoring. It is usually solved in following way:
* take one task at a time, something distinct that causes trouble, that you want to refactor
* investigate the code to learn how it works, focusing only on the chosen "thing" (ignore anything else)
* discuss it with other team members, find available solution, and set course of action
* work only on that task until its done, then merge.

For example, in the past I tried to investigate that resolution/scaling issue you were talking about. From my findings, it seems that AGS is doing conversion between script and game coordinates. The problem is that it does so in erratic and inconsistent way: some of the game variables keep coordinates in "game" values, and others keep them in "script" values.

The solution that comes to mind is following: move and restrict all coordinate conversion to the place where "script" connects with the "game". Luckily, we already have such place, because of refactoring we did years ago. This is the special functions registered in script systems, the ones that look like:
<pre>RuntimeScriptValue Sc_DrawingSurface_CreateCopy(void *self, const RuntimeScriptValue *params, int32_t param_count)</pre>

This is a best place to put all these of conversions. Basically, if a script function receives coordinates in parameters, you convert these parameters into game coord system. And if a script function returns coordinate, you convert it back. Changing weird name of a function from **multiply_up_coordinate** to **convert_script_to_game_coord** would also improve things, making it easier to understand what is going on.

I am giving this example just to point out that such things are doable.



***

@AlanDrake , you idea was to make the code better and suitable for further development. And your idea was to simpy remove all the legacy functionality (or rather the obvious part of it) at once.

So, as you see in my previous comment, my biggest fear is that the program will be left in a half-working state with no one to maintain it.

A year ago, when you just started this discussion, I was not in a very good state to participate or help, mostly because there was too much on my hands (version 3.4.0 had to be finished, and also we were developing 3.4.1 at the same time).

What has changed since is, that the actual development has currently slowed down, and I do not have any big plans about it anymore. There are certain bugs that still need to be fixed, but that's a minor thing.

So, I've been thinking ... really hard, trying to find a way out of this...
And I would like to suggest an alternative way.

What if, **instead of just deleting all the old code**, we would start a **big refactoring project**.

The rules of this "project" are:

1) We create "classic" or "stable" branch at current master, this will be the branch to complete AGS 3.4.2, and maybe create 3.5.0 with SDL support from @sonneveld .

2) Master branch becomes the place of a grand refactoring project.

3) It is made absolutely clear, than master branch is not guaranteed to be safe to use until we say so (come to certain stage in refactoring). We allow it to break things from time to time, and don't pursue 100% functionality until later stages.

4) No new features are allowed, maybe with few exceptions that don't counteract refactoring.

5) We work altogether, refactoring **part by part.** The less "part" is, the better (when possible).

6) We discuss refactoring, each task in separate issue ticket. This way everyone knows the problem in question, and can help and give advices.


What do you think about this?

**Comment by AlanDrake** [comment]
I don't care for refactoring legacy code, I think it's a waste of time.
The only reason I ever picked up AGS in the first place was for its ability to make adventure games. The thought it could run older games never once crossed my mind.
I don't want to sound harsh, but 99% of the potential game developers out there, only cares about making new games.

You know I started this whole cleanup for making it easier to refactor and evolve, so I really don't see why you're now suggesting a big refactor on master. It's more work and more frustrating.
Did anything change to make you think it's possible now, when it wasn't possible three years ago ?

I feel it would be a mistake. We just don't have the manpower and I don't want to do it. I've made it clear various times, I don't want to deal with legacy code. 
So either we cut away the dead weight and work together to clean, refactor and improve AGS, or I'm throwing the towel.

**Comment by ivan-mogilko** [comment]
So, what should be the plan then? We declare that master is under big rewrite and should not be trusted.

There is a lot of things that probably should be scrapped along with the legacy code. Everything that imposes constraints on data structures, making it unable to modify them. That is: plugin API, the memory variables directly exposed into script API.

**Comment by ivan-mogilko** [comment]
@sonneveld ,
> Leaving legacy code in its own branch will mean it will wither and die. The only alternative to run old games on new hardware is to hope that scummvm supports it.

I agree with this, but as you see, this project was from the start suffering from unclear goals and mixed expectations. When I joined, actually I wanted to do about same thing as @AlanDrake is doing now. That was also the reason for me to join. I knew that Alyssa Milburn from scummVM team is working on a port, so I figured that such port will make old games covered.

There were bad decisions made, and we started a new development over JJS's ports, which goals, in turn, was to support all the old games. ScummVM port was never finished, and we got stuck in what we have.
To tell the truth, I always felt terrible about this, because I too had a lot of plans regarding AGS, which I could never bring to life, and I also never intended to be maintainer of a program to run legacy games. If I could return back in time, certainly I would choose another path.

Ideally we probably should be doing what you say, but it's true that there is not enough people involved, and there is also the fact that too much time was lost. I am not even sure if AGS will be used a lot in the future.

It is still probably possible to kind of mix Alan's solution with yours. For example, when the new cleaned branch is set and tidied more or less, we may begin bringing old support piece by piece, in such way that won't trouble current development.

**Comment by AlanDrake** [comment]
Yes, we declare that we dropped legacy support and are in the process of a rewrite. Baseline for importing projects will be minimum the latest 3.4.1 until a new game format is decided. I'll write a few notes about the quirks related to importing projects.

At the moment the current ags4 branch should be able to run modern ags games without issues, except in cases where sprites were flagged as *lowres* and the game was *hires*, or a deprecated functionality was used.

I've been adding "// CLNUP" comments to potential places where to intervene. I'm counting on your expertise to figure out other points needing to be addressed.
Some code has been commented and could be removed, if not needed for reference.

As how to move toward the future I'm not sure. 

Here's the list of changes:
* Removed UseLowResCoordinatesInScript option, highResCoordinates, 
* Removed everything related to LowRes/HiRes sprite flags, that required multiplying or dividing sizes/coords in many places
* Removed "resolution" property from rooms (as in "high" or "low", not actual size)
* Stretch masks to room resolution if needed (previously scaled according to "resolution")
* Removed operator precedence option
* Removed save/load routines for versions < 3.4.1
* Removed GameSpriteAlphaRenderingStyle and GameGuiAlphaRenderingStyle options
* Removed CD Audio support
* Removed old interaction system
* Removed legacy dialog support
* Removed various deprecated scripting functions (incomplete, tracked on the "battle plan" google doc)

Other stuff removed was unused variables and dead code.

**Comment by necros2k7** [comment]
Ok, deleting old games support, but how can people run old ac2games/DOS games? Calsoon for example, Crime time, Book of spells? Isn`t it a good idea to support all engine versions in one runtime exe? (like for example ScummVM supporting wide versions of SCI , AGI engines for example). I`m ok with engine cleaning but if there is a standalone win compatible legacy DOS AGS engine gamers runner

**Comment by AlanDrake** [comment]
Ideally, the legacy game engine itself should go in a separate module.
That way it would make possible to support both old and new development.
The problem is untangling the engine framework from the game engine logic, so that they become indepentent from each other.
Once 3.5 is finalized and merged, I'm planning to attempt this for AGS4 in order to see how it could be done. If succeeded, it may be used as a blueprint to perform the same for AGS Classic.

In the end what we should be aiming for would be to get the following components:
* Framework engine (I/O, rendering, sound, etc)
* New game engine
* New project format
* Legacy game engine (2.x and 3.x, which could be split apart if it improves maintenance)
* Project importer/converter for older formats

The pros:
* Easier to work on compatibility improvements because the framework engine is stand-alone
* Less dangers of causing game logic regressions
* The new engine can be developed in parallel without impacting or breaking older games
* The possibility of building an AGS player for specific versions
* Easier to add new features and making big changes

The cons:
* Needs a solid plan for the framework engine so no regrets arise later
* It will take time and effort because AGS codebase is not a small one

How to get there though? How to design the framework API?
My main focus is making a game not making game engines, we'll need to work together to make this happen.

**Comment by fernewelten** [comment]
> Ok, deleting old games support, but how can people run old ac2games/DOS games? 

One obvious solution would be to continue using the existing engines that the games were designed for and bundled with. The old executables don't drop off the ends of the world just because new versions are being released.

I don't think it is quite fair to compare the AGS to the ScummVM project. 

The original SCUMM system was developed for the specific respective current game it was used for, and it didn't have compatibility of former games in mind. As far as I know, nobody tried to keep the SCUMM engine compatible to former, already published games, when a new game was being developed and that current game had specific new requirements. This is one of the reasons that the ScummVM project needs to deal with lots of one-shot anomalies, copy protections, etc.

The ScummVM on the other hand has been designed from the very get-go as a historic emulator project, after the demise of the Lucas Arts / Sierra studios. They painstakingly reconstruct copy protection mechanisms, one-shot kludges. and curious quirks of the ancient games, ideally in order to get them all to run.

In a similar vein, an "AGSVM" emulator could be maintained in a separate project to make ancient AGS executables continue to run. Developers that are into that kind of thing (I'm told there are some) can disassemble bytecode and find out how things used to be done in yonder years and olden times. More so, they can even access the olden sources, due to the wonders of code version systems.

On the other hand, the AGS has been designed right from the get go in Chris Jones' times as a **tool for developing new adventure games**, just as the original SCUMM system was. Should compatibility to ancient versions impede progress seriously, it can drop compatibility and move on.

It is debatable whether this is **indeed** the case. As far as I understand, @AlanDrake 's argument is that this point has been reached. His point is that compatibility concerns have made the code so complicated that it has become inordinately difficult to implement new features and to hunt down bugs. The code can be made much sleeker and more maintainable if those compatibility features are all dropped that nobody uses any more in new games anyway.

If that is so, then that ballast should be jettisoned to re-gain the ability to make progress. Compatibility to old, already compiled executables would be a secondary concern.

**Comment by fernewelten** [comment]
I fear that this will become a moot point very soon, because AFAICS, **nobody** is doing anything non-trivial in the AGS4 branch, nor have they in the last year or so; and **nobody** whatsoever has championed the branch in the last year. All the efforts have been directed to just keeping the branch fairly in sync with AGS 3.5, with @ivan-mogilko being the only one saddled with that work. (I might be unfairly biased in this because I didn't really keep close tabs.)

Without any champions and without any momentum, this branch will be quickly overtaken by the power of the factual and left behind. 

To make the branch thrive, developers will need good reasons to want to switch, so the branch will need to offer them advantages that can't be had in the 3.5 branch. And end users will need good reasons to prefer an "4.0 AGS" over an "3.5 AGS", so just taking out deprecated code isn't enough: Something new would need to be put in that would make "4.0 AGS" seem superior from a user perspective. But as far as I know, the 4.0 branch hasn't even compiled to a complete studio in the last three months or so so that end users can't test and compare it to the "3.5 AGS".

**Comment by ivan-mogilko** [comment]
> nobody is doing anything non-trivial in the AGS4 branch, nor have they in the last year or so; and nobody whatsoever has championed the branch in the last year. All the efforts have been directed to just keeping the branch fairly in sync with AGS 3.5

To be fair, early last year I was not planning to continue development in ags3 branch at all and even started doing some further changes (stripping on legacy functionality) in ags4. The reason 3.5 exists as 3.5 was basically that there were continious "urgent" user requests regarding things that were "spoling their lives" in ags. As ags4 could still be kind of unstable after all the things ripped of (and ofc because user projects could still use old stuff) I decided to make "one last version" with ags3 code.

**Comment by AlanDrake** [comment]
Speaking of which, how close is the SDL port to completion?
It was a big reason why I never attempted anything too drastic. I didn't want to make the branch unmergeable.

**Comment by ivan-mogilko** [comment]
I think to close this thread, it has been here for 3 years, last updated year ago, and guess everything that had to be said is already said, so...
